src/swap.ts

```diff
@@ -0,0 +1,291 @@
+import { purry } from './purry';
+import { isString } from './isString';
+import { isArray } from './isArray';
+import { IterableContainer } from './_types';
+
+type isEqual<A, B> = A extends B ? (B extends A ? true : false) : false;
+
+type isLessThan<A extends number, B extends number> = isEqual<A, B> extends true
+  ? false
+  : 0 extends A
+  ? true
+  : 0 extends B
+  ? false
+  : isLessThan<Subtract<A, 1>, Subtract<B, 1>>;
+
+type BuildTuple<
+  L extends number,
+  T extends ReadonlyArray<unknown> = []
+> = T['length'] extends L ? T : BuildTuple<L, [...T, unknown]>;
+
+type Subtract<A extends number, B extends number> = BuildTuple<A> extends [
+  ...infer U,
+  ...BuildTuple<B>
+]
+  ? U['length']
+  : never;
+
+type IsNonNegativeInteger<T extends number> = number extends T
+  ? false
+  : `${T}` extends `-${string}` | `${string}.${string}`
+  ? false
+  : true;
+
+type Absolute<T extends number> = `${T}` extends `-${infer R}`
+  ? R extends `${infer N extends number}`
+    ? N
+    : never
+  : T;
+
+type StringToChars<T extends string> = string extends T
+  ? Array<string>
+  : T extends `${infer C}${infer R}`
+  ? [C, ...StringToChars<R>]
+  : [];
+
+type Join<
+  Strings extends ReadonlyArray<string>,
+  Accumulator extends string = ''
+> = Strings extends readonly [infer Head, ...infer Rest]
+  ? Rest extends ReadonlyArray<string>
+    ? Join<
+        Rest,
+        Head extends string
+          ? Accumulator extends ''
+            ? Head
+            : `${Accumulator}${Head}`
+          : never
+      >
+    : Accumulator
+  : Accumulator;
+
+type ObjectSwap<T, K1 extends keyof T, K2 extends keyof T> = {
+  [K in keyof T]: T[K1 extends K ? K2 : K2 extends K ? K1 : K];
+};
+
+type SwapArrayInternal<
+  T extends IterableContainer,
+  Index1 extends number,
+  Index2 extends number,
+  Position extends IterableContainer = [],
+  Original extends ReadonlyArray<unknown> = T
+> = number extends Index1
+  ? T
+  : number extends Index2
+  ? T
+  : T extends readonly [infer AtPosition, ...infer Rest]
+  ? [
+      Position['length'] extends Index1
+        ? Original[Index2]
+        : Position['length'] extends Index2
+        ? Original[Index1]
+        : AtPosition,
+      ...SwapArrayInternal<
+        Rest,
+        Index1,
+        Index2,
+        [unknown, ...Position],
+        Original
+      >
+    ]
+  : T;
+
+type SafeNegativeIndex<
+  T extends IterableContainer,
+  Index extends number
+> = Subtract<T['length'], Absolute<Index>>;
+
+type SafePositiveIndex<
+  T extends IterableContainer,
+  Index extends number
+> = isLessThan<Index, T['length']> extends true ? Index : never;
+
+type SafeIndex<
+  T extends IterableContainer,
+  Index extends number
+> = IsNonNegativeInteger<Index> extends false
+  ? SafeNegativeIndex<T, Index>
+  : SafePositiveIndex<T, Index>;
```

Comment:
I'd flip the order here..
```suggestion
> = IsNonNegativeInteger<Index> extends true
  ? SafePositiveIndex<T, Index>
  : SafeNegativeIndex<T, Index>;
```